src/router.rs

```diff
@@ -125,46 +126,101 @@ impl RequestForwarder {
         Ok(())
     }
 
-    #[allow(clippy::single_match_else)]
-    pub fn forward_request(&self, request: ComponentRequest) -> Result<Response<Body>, RouterError> {
-        let component_name = format!("{}/{}", request.user, request.repo);
-        //if !self.components_map.read().contains_key(&component_name) {
-        //    self.update_workers()?;
-        //}
-
-        // TODO: Make this more efficient (when do I actually need to call update_workers()?)
-        self.update_workers()?;
-
-        debug!("Forwarding request to: {:?}", component_name);
-        let locked_map = self.components_map.read();
-        let target_component_option = locked_map.get(&component_name);
-
-        let target_component = match target_component_option {
-            Some(v) => v,
-            None => {
-                let body = Body::from("Requested component not found\n");
-                warn!("Component to forward to not found: {:?}", component_name);
-                return Ok(Response::builder().status(404).body(body).unwrap());
-            }
-        };
+    fn find_appropriate_worker(&self, component_name: &str) -> Option<Arc<WorkerNode>> {
+        // We want to keep the happy path free of lock contention, so we clone the Arc<WorkerNode>
+        // (Otherwise we'd have to keep a read lock for the duration of each call to the component, which
+        // could cause unnecessary contention for other threads trying to call `update_workers`)
+        self.components_map.read().get(component_name).cloned()
+    }
 
-        let mut url = format!(
-            "{}/sl/{}/{}",
-            target_component.url, component_name, request.method
-        );
+    fn send_request_to_worker(
+        request: ComponentRequest,
+        component_name: &str,
+        worker_url: &str,
+    ) -> Result<Response<Body>, RouterError> {
+        let mut url = format!("{}/sl/{}/{}", worker_url, component_name, request.method);
 
         if !request.query.is_empty() {
             url = format!("{}?{}", url, request.query);
         }
 
+        // TODO: This blocks the executor, so we probably should do something smarter than just blocking
```

Comment:
This won't scale to different connections at a time, will it? If not, we definitely need to find a better way

Reply:
I think by default the executor has #threads = #cores. (And this only blocks one thread [I think]) So it isn't as bad as being single threaded. The right way to do this is probably to actually use tokio directly/make everything in an async context.

I want to try this in the worker first, so let's hold off fixing for now